test/chain-tree-compaction-test.js

```diff
@@ -0,0 +1,581 @@
+
+        // Current tree root before crash
+        const treeRoot = chain.db.treeRoot();
+
+        // Implement bug where node crashes before database batch is written.
+        // When the next block is connected, it should successfully write
+        // new data to the Urkel Tree but fail to write data to blockstore
+        // or levelDB indexes.
+        chain.db.commit = async () => {
+          // Tree root has been updated inside Urkel
+          const newRoot1 = chain.db.treeRoot();
+          assert(!treeRoot.equals(newRoot1));
+
+          // Reset batch, otherwise assert(!this.current) fails
+          chain.db.drop();
+          // Node has crashed...
+          await chain.close();
+        };
+
+        // Update name and attempt to confirm
+        send(update, mempool);
+        // Will "crash" node before completing operation
+        await mineBlocks(1, mempool);
+        assert(!chain.opened);
+
+        // Restore proper batch-write function
+        chain.db.commit = CHAIN_DB_COMMIT;
+
+        // Restarting chain should recover from crash
+        await chain.open();
+
+        // Tree root has been restored from pre-crash state
+        const newRoot2 = chain.db.treeRoot();
+        assert(treeRoot.equals(newRoot2));
+
+        // Try that update again with healthy chainDB
+        send(update, mempool);
+        await mineBlocks(1, mempool);
+
+        // Tree has been updated but tree root won't be committed
+        // to a block header until the next block.
+        assert(!chain.db.tree.rootHash().equals(chain.tip.treeRoot));
+        await mineBlocks(1);
+
+        // Everything is in order
+        assert.bufferEqual(chain.db.tree.rootHash(), chain.tip.treeRoot);
+        raw = await chain.db.tree.get(nameHash);
+        ns = NameState.decode(raw);
+        assert.bufferEqual(ns.data, Buffer.from([counter + 1]));
+      });
+
+      it(`should ${prune ? '' : 'not '}have pruned chain`, async () => {
+        // Sanity check. Everything worked on a chain that is indeed pruning.
+        // Start at height 2 because pruneAfterHeight == 1
+        for (let i = 2; i <= chain.height; i++) {
+          const entry = await chain.getEntry(i);
+          const block = await chain.getBlock(entry.hash);
+
+          if (prune && i <= chain.height - network.block.keepBlocks)
+            assert.strictEqual(block, null);
+          else
+            assert(block);
+        }
+      });
+    });
+  }
+
+  describe('SPV', function() {
+    it('should ignore compact tree option', async () => {
+      const prefix = path.join(
+        os.tmpdir(),
+        `hsd-tree-compacting-test-${Date.now()}`
+      );
+
+      const node = new SPVNode({
+        prefix,
+        network: 'regtest',
+        memory: false,
+        compactTree: true,
+        prune: true // also ignored
+      });
+
+      await node.ensure();
+      await node.open();
+      await node.close();
+    });
+  });
+
+  describe('Full Node', function() {
```

Comment:
Same as in SPV.

Reply:
5420112e6c7cd7faf14f3f9e446f8360ef83cc6b